lib/index.js

```diff
@@ -1,3 +1,88 @@
 'use strict';
 
-module.exports = {};
+const { Kinesis: AwsKinesis } = require('aws-sdk');
+const { PassThrough } = require('stream');
+const ShardSubscriber = require('./shard-subscriber');
+const consumer = require('./consumer');
+const stream = require('./stream');
+const { noop } = require('./utils');
+
+const privateData = new WeakMap();
+
+function internal(instance) {
+  if (!privateData.has(instance)) privateData.set(instance, {});
+  return privateData.get(instance);
+}
+
+class Kinesis extends PassThrough {
+  constructor(options = {}) {
+    super({ objectMode: true });
+    const {
+      compression,
+      consumerName,
+      createStreamIfNeeded = true,
+      encryption,
+      logger = {},
+      shardCount = 1,
+      streamName,
+      tags,
+      ...otherOptions
+    } = options;
+
+    const normLogger = {
+      debug: typeof logger.debug === 'function' ? logger.debug.bind(logger) : noop,
+      error: typeof logger.error === 'function' ? logger.error.bind(logger) : noop,
+      warn: typeof logger.warn === 'function' ? logger.warn.bind(logger) : noop
+    };
+
+    if (!consumerName) {
+      const errorMsg = 'The "consumerName" option is required.';
+      normLogger.error(errorMsg);
+      throw new TypeError(errorMsg);
+    }
+
+    if (!streamName) {
+      const errorMsg = 'The "streamName" option is required.';
+      normLogger.error(errorMsg);
+      throw new TypeError(errorMsg);
+    }
+
+    Object.assign(internal(this), {
+      compression,
+      consumerName,
+      createStreamIfNeeded,
+      encryption,
+      logger: normLogger,
+      shardCount,
+      streamName,
+      tags,
+      options: otherOptions
+    });
+  }
+
+  async connect() {
+    const ctx = internal(this);
```

Comment:
I'd be in favor of calling this "context". I was confused by the meaning of "ctx" when I first saw this...